Add Protobuf and ODE as ExternalProjects to cmake to fix platform compatibility issues

owlkward opened the pull request:
Builds Protobuf from source if the version is too new.
I'm not sure which protobuf version was the first to break, but if someone notices an incompatibility in the future they  can just adjust the maximum allowed version.

For ODE I didn't find a way to automatically detect incompatibility, because it doesn't seem to be completely dependent on the version number, but could also depend on build configurations of the package, like enabling double precision or not.
e.g. if double  precision is not enabled grSim crashes instantly, but with a different error than when using the Manjaro ode package.
So it ended up as an optional flag to enable building ODE from source.
Personally I'd make building ODE from source the default, but that seems more invasive than keeping the current behavior as default.

I don't know how GitHub Actions works, so I didn't test whether the pipeline works, yet.

g3force commented:
For GitHub Actions, you can configure the build here: https://github.com/RoboCup-SSL/grSim/blob/master/.github/workflows/build.yaml#L33

You can remove ode and protobuf from brew installation and instead set the cmake variable to use external ODE.

owlkward commented:
macos build should be fine now, I hope.
I also tried to fix the failing windows build by telling cmake where to look for the Qt5 it apparently isn't finding, but I'm not sure if that actually works to just add the vcpkg to CMAKE_PREFIX_PATH or if it needs to be the specific path to the Qt5 installation.

owlkward commented:
nevermind, I was wrong, the dependencies on windows were not actually ever installed? Fixed that as well, I think.
Maybe windows works now, but I don't know yet, because the pipeline has been running for an hour, because apparently vcpkg compiles all dependencies from source and the dependency listed qt5 instead of qt5-base (I also changed that now, but I'm not going to wait until it finishes, I need sleep), which takes hours to compile even on a good machine.

macos now also definitely works (pipeline on my fork is successful for both macos versions), but it includes a hack for the PATH to the Qt5Config.cmake, which I'm not happy about, I have no idea on how paths on macos works and especially not on macos inside a github runner.

owlkward commented:
Apart from the stuff that should in theory fix the windows pipeline, but doesn't actually fix it, ODE now also gets built from source if the flag is not set, but only if find_package can't find it.

owlkward commented:
oh and this pull request should fix #176 and make https://github.com/RoboCup-SSL/grSim/pull/179 unnecessary
